dll/win32/shell32/utils.cpp

```diff
@@ -540,6 +540,123 @@ LargeIntegerToString(
                          pNumberFormat, dwNumberFlags);
 }
 
+/*************************************************************************
+ *  CopyStreamUI [SHELL32.726]
+ *
+ * Copy a stream to another stream with optional progress display.
+ */
+EXTERN_C
+HRESULT WINAPI
+CopyStreamUI(
+    _In_ IStream *pSrc,
+    _Out_ IStream *pDst,
+    _Inout_opt_ IProgressDialog *pProgress,
+    _In_opt_ DWORDLONG dwlSize)
+{
+    HRESULT hr = E_FAIL;
+    DWORD cbBuff, cbRead, dwSizeToWrite;
+    DWORDLONG cbDone;
+    LPVOID pBuff;
+    STATSTG Stat;
+    BYTE szBuff[1024];
+
+    TRACE("(%p, %p, %p, %I64u)\n", pSrc, pDst, pProgress, dwlSize);
+
+    if (dwlSize == 0) // Invalid size?
+    {
+        // Get the stream size
+        ZeroMemory(&Stat, sizeof(Stat));
+        if (FAILED(pSrc->Stat(&Stat, 1)))
+            pProgress = NULL; // No size info. Disable progress
+        else
+            dwlSize = Stat.cbSize.QuadPart;
+    }
+
+    if (!pProgress) // Progress is disabled?
+    {
+        ULARGE_INTEGER uliSize;
+
+        if (dwlSize > 0)
+            uliSize.QuadPart = dwlSize;
+        else
+            uliSize.HighPart = uliSize.LowPart = INVALID_FILE_SIZE;
+
+        return pSrc->CopyTo(pDst, uliSize, NULL, NULL); // One punch
+    }
+
+    // Allocate the buffer if necessary
+    if (dwlSize > 0 && dwlSize <= sizeof(szBuff))
+    {
+        cbBuff = sizeof(szBuff);
+        pBuff = szBuff;
+    }
+    else
+    {
+#define COPY_STREAM_DEFAULT_BUFFER_SIZE 0x4000
+        cbBuff = COPY_STREAM_DEFAULT_BUFFER_SIZE;
+        pBuff = LocalAlloc(LMEM_FIXED, cbBuff);
+        if (!pBuff) // Low memory?
+        {
+            cbBuff = sizeof(szBuff);
+            pBuff = szBuff;
+        }
+#undef COPY_STREAM_DEFAULT_BUFFER_SIZE
+    }
+
+    // Start reading
+    LARGE_INTEGER zero;
+    zero.QuadPart = 0;
+    pSrc->Seek(zero, 0, NULL);
+    pDst->Seek(zero, 0, NULL);
+    cbDone = 0;
+    if (pProgress) // Progress is enabled?
+        pProgress->SetProgress64(cbDone, dwlSize);
+
+    // Repeat reading and writing until goal
+    while (SUCCEEDED(pSrc->Read(pBuff, cbBuff, &cbRead)))
+    {
+        // Calculate the size to write
+        if (dwlSize > 0)
+            dwSizeToWrite = (DWORD)min((DWORDLONG)(dwlSize - cbDone), (DWORDLONG)cbRead);
+        else
+            dwSizeToWrite = cbRead;
+
+        if (dwSizeToWrite == 0) // No need to write?
+        {
+            hr = S_OK;
+            break;
+        }
+
+        hr = pDst->Write(pBuff, dwSizeToWrite, NULL);
+        if (hr != S_OK)
+            break;
+
+        cbDone += dwSizeToWrite;
+
+        if (pProgress) // Progress is enabled?
```

Comment:
There is an early bail-out if pProgress is null, why the check here?

Reply:
Removed check.